Update the build system to support installing.

cnweaver opened the pull request:
Libraries and python scripts are installed to the standard python module location, and headers and cmake configuration scripts are installed to the selected prefix directory.

nwhitehorn commented:
Resolves #21

mhasself commented:
I am thrilled to see this!  I will take a closer look this week but there a couple of things that I'm immediately wondering about:
- Can one override the Python package install prefix at time of cmake invocation?
- Perhaps the README and quickstart should be updated with this new business?

arahlin commented on cmake/Spt3gConfig.cmake.in:
```diff
@@ -1,14 +1,28 @@
+	set(SPT3G_INCLUDE_DIRS "@PACKAGE_INCLUDE_INSTALL_DIR@/include/core")
+	foreach(lib @SPT3G_LIBRARIES@)
+		set(SPT3G_LIBRARIES ${SPT3G_LIBRARIES} spt3g::${lib})
+		set(SPT3G_INCLUDE_DIRS "@PACKAGE_INCLUDE_INSTALL_DIR@/${lib}")
```
Is `SPT3G_INCLUDE_DIRS` supposed to be a growing list of directories?  Looks like this might be a bug that's propagated from before.

cnweaver replied:
I think it is; in general `find_package(X)` produces variables `X_INCLUDE_DIRS` and `X_LIBRARIES` which downstream users can feed into `target_include_directories` and `target_link_libraries`, basically to set the appropriate `-I` and `-l` flags for building things which depend on X.

arahlin replied:
Yeah, my point is that that line should probably be this I think (currently missing the original list variable to append to):

```
set(SPT3G_INCLUDE_DIRS ${SPT3G_INCLUDE_DIRS} …)
```

cnweaver replied:
Oh, yes! I will fix that. Sorry; got distracted from the trees by the forest.

cnweaver replied:
This should now be fixed, as is the test for whether the installed or uninstalled mode should be used.

arahlin reviewed (changes requested):
This looks great!  I haven't tried running it yet, but will do in the next couple days.  One minor nit I noticed while looking through the changes so far.  Also, agreed with @mhasself's comments

arahlin commented:
Another thought: it would be convenient to make the install work with a `setup.py` file so that down the road we can make the package pip-installable.   I looked into this at one point and I think it’s doable to add build extensions that run cmake to make this work, not sure how tricky it would actually be to implement though.

cnweaver commented:
The python package install directory can now be overridden, and that and the rest of this feature are now documented. 

It should certainly be possible to write a setuptools script to drive the cmake build process as I think I have seen done elsewhere, but I have no experience with how to do it. I'm not sure how practical installation through `pip` would actually be, given the need for dependencies not installed through pip (e.g. boost), but if there's a will to do it I'm sure it can be done.

arahlin commented:
Just tried installing the package and it works great.  However, I'm wondering if we want to rename the installed libraries that end up in `/usr/local/lib` to `libspt3gcore`, `libspt3gmaps` etc... as is, those names aren't super easy to search.  It looks like separate copies of the libraries end up in the python package anyway.  Maybe some more thought into how best to deal with linking against the installed libraries/package is needed?  I know the SO crew has put in some work on importing the C++-built python bindings in a smarter way that might be useful to integrate here, or as a separate PR.

arahlin commented:
A few other things I noticed:

* There are some cmake syntax warnings with the cmake_uninstall file, which I think can be resolved by doing the message string construction with a single set of quotes rather than concatenating strings and variables together.  Might be an issue with newer cmake version (I am running 3.19.3 locally).
* The uninstaller doesn't remove the include/spt3g, share/spt3g and site-packages/spt3g directories when it's done.  I think it should probably remove those too?

cnweaver commented:
The warnings due to bad quoting were also present with cmake 3.20; I had just failed to notice them because `make install` produces a large aount of output. They should now be fixed. 

The directories were being left behind because cmake was producing them as a by-product of putting files into them, and apparently not adding a record of having done so to the install manifest. There is now some special-case code to clean them up. 

> I'm wondering if we want to rename the installed libraries that end up in /usr/local/lib to libspt3gcore, libspt3gmaps etc... as is, those names aren't super easy to search.

I would like to do this, as the naming is a bit strange at the moment, since it generally obeys neither standard library naming nor python naming, although apparently all uses are flexible enough that it's actually working. `set(CMAKE_SHARED_LIBRARY_PREFIX "")` is what's causing the weird naming, but since this is somewhat visible from inside of python, I don't want to touch it. Unfortunately, CMake does not support renaming targets during the install process (and `install(TARGET)` must be used for its `EXPORT` option which has the necessary side effect of priming the CMake module information for the later `install(EXPORT)`), and it is also unclear what impact this might have on code buildtagainst `libfoo.(so|dylib)` when used from within python with `foo.(so|dylib)` having been loaded instead, or with  the exported cmake targets referring only to the un-prefixed names. Ultimately, this seems to just be a disadvantage of the peculiar structure of bundling the python and non-python portions into single libraries.

arahlin commented:
Maybe @mhasself can comment on this, since he’s likely to be one of first people to use this feature, but I wonder if we want to maybe try to follow the numpy package model and install the include and share directories directly inside the python package?  That would bypass both the issues with the installed library names and the duplicate libraries.  We could even include convenience python functions for finding the header / library / cmake file directories for linking.

cnweaver commented:
Personally, I consider the numpy method of doing this rather awful to deal with, and I don't know that it would interact very well with cmake exported targets. In particular, I'm not aware of a natural place to put the cmake module definitions within python's subdirectory structure, nor would cmake know to search there in the way that it does automatically when everything uses a common install prefix in the standard non-python way. The current state of this patch does work with so3g's existing build system, which was the thing which put me on the path to working on it.

arahlin commented:
That's fair.  I guess this version is fine for now, would be great to handle the libraries properly at some point down the line.

mhasself commented:
> > I'm wondering if we want to rename the installed libraries that end up in /usr/local/lib to libspt3gcore, libspt3gmaps etc... as is, those names aren't super easy to search.
> 
> I would like to do this, as the naming is a bit strange at the moment, since it generally obeys neither standard library naming nor python naming, although apparently all uses are flexible enough that it's actually working. `set(CMAKE_SHARED_LIBRARY_PREFIX "")` is what's causing the weird naming, but since this is somewhat visible from inside of python, I don't want to touch it. Unfortunately, CMake does not support renaming targets during the install process (and `install(TARGET)` must be used for its `EXPORT` option which has the necessary side effect of priming the CMake module information for the later `install(EXPORT)`), and it is also unclear what impact this might have on code buildtagainst `libfoo.(so|dylib)` when used from within python with `foo.(so|dylib)` having been loaded instead, or with the exported cmake targets referring only to the un-prefixed names. Ultimately, this seems to just be a disadvantage of the peculiar structure of bundling the python and non-python portions into single libraries.

In `so3g`, I balked at the lib-less library names and so rewrote things a bit to produce libso3g.(so|dylib).  The python symbols are dragged into the main module in the same way as `spt3g.*`; see so3g's [__init__.py](https://github.com/simonsobs/so3g/blob/90f4679d5dc7b26b61b4329ae1edc1320c624270/python/__init__.py#L12) and [our version of load_bindings](https://github.com/simonsobs/so3g/blob/90f4679d5dc7b26b61b4329ae1edc1320c624270/python/load_pybindings.py#L15).  It works fine.  (And like with spt3g.*, the classes present clean fundamental names ... e.g., so3g.RangesInt32, not so3g.libso3g.RangesInt32 or so3g._so3g.RangesInt32...)

nwhitehorn commented:
> > > I'm wondering if we want to rename the installed libraries that end up in /usr/local/lib to libspt3gcore, libspt3gmaps etc... as is, those names aren't super easy to search.
> > 
> > 
> > I would like to do this, as the naming is a bit strange at the moment, since it generally obeys neither standard library naming nor python naming, although apparently all uses are flexible enough that it's actually working. `set(CMAKE_SHARED_LIBRARY_PREFIX "")` is what's causing the weird naming, but since this is somewhat visible from inside of python, I don't want to touch it. Unfortunately, CMake does not support renaming targets during the install process (and `install(TARGET)` must be used for its `EXPORT` option which has the necessary side effect of priming the CMake module information for the later `install(EXPORT)`), and it is also unclear what impact this might have on code buildtagainst `libfoo.(so|dylib)` when used from within python with `foo.(so|dylib)` having been loaded instead, or with the exported cmake targets referring only to the un-prefixed names. Ultimately, this seems to just be a disadvantage of the peculiar structure of bundling the python and non-python portions into single libraries.
> 
> In `so3g`, I balked at the lib-less library names and so rewrote things a bit to produce libso3g.(so|dylib). The python symbols are dragged into the main module in the same way as `spt3g.*`; see so3g's [**init**.py](https://github.com/simonsobs/so3g/blob/90f4679d5dc7b26b61b4329ae1edc1320c624270/python/__init__.py#L12) and [our version of load_bindings](https://github.com/simonsobs/so3g/blob/90f4679d5dc7b26b61b4329ae1edc1320c624270/python/load_pybindings.py#L15). It works fine. (And like with spt3g.*, the classes present clean fundamental names ... e.g., so3g.RangesInt32, not so3g.libso3g.RangesInt32 or so3g._so3g.RangesInt32...)

I think that's totally fine. The lib-less names were meant to allow purely-C++ modules to be imported naturally into Python. In point of fact, we don't have any purely-C++ modules, so that motivation has passed its expiration date.

mhasself commented:
After a bunch of wrestling (between conda and arch, where they took a lot out on the ref), I have been able to test this.  I was pleased that (in the end), so3g built without any trouble.

My only issue that prevents this from being strictly better than what I was doing before is that I don't know how to build-and-install only a subset of the spt3g modules.  I only really use core, so I usually run cmake, and then `make core version`... and then I have my own terrible install scripts to copy the result out.

In the present system, the "make install" requires all discovered modules to be built...  Perhaps there's a cmake trick to only configure core?

You must understand... I really, really don't need gcp.so! :)

cnweaver commented:
Being able to build only a subset of modules should definitely be doable; I've seen something like that in other projects which use CMake. I think it can probably be just as simple as a `MODULES_TO_BUILD` variable that the user can set, which is then used to filter the available modules whose source code is actually found, as long as everyone is okay with it not really being able to evaluate dependencies. That is, if module `foo` requires module `bar`, but the user requests only `foo`, it might fail with a non-optimal explanation about why. It would probably make sense to at least automatically add `core` to the module list if it isn't specified, though, since I doubt anything else works without it.

cnweaver commented:
Building a subset of libraries is now implemented, as is putting back the standard library prefix. I found that copying the `spt3g` subdirectory from the build directory was resulting in installing a bunch of symlinks instead of the actual python scripts, but more complex code was required to sort that out, thanks to some weird omissions from cmake.

arahlin commented on CMakeLists.txt:
```diff
@@ -157,14 +173,31 @@ if(NOT PYTHON_MODULE_DIR)
+		continue() # skip phony library
+	endif("${lib}" STREQUAL "spt3g")
+
+	if(EXISTS ${CMAKE_SOURCE_DIR}/${lib}/python)
```
Some of the libraries in the private repository don’t have a `python` us directory.  There should be a function that sets that directory and creates the symlinks, maybe the install code can be part of that function instead?

mhasself commented:
Thanks -- just tested, and this is working almost perfectly for my purposes...

For me, "make install" does not copy over spt3g/version.py, so package import fails... but once I copy that one file over manually, it's fine.  So this is very close now!

(Here's my cmake line if that's helpful:
```
cmake -DPYTHON_MODULE_DIR=${DEST}/site-packages -DCMAKE_INSTALL_PREFIX=${DEST} -DBUILD_PROJECTS='core' ..
```

(I've also tried specifying 'core;version' -- the problem is not that version.py is not built, it's that it is not installed.)

cnweaver commented:
> Some of the libraries in the private repository don’t have a python subdirectory. There should be a function that sets that directory and creates the symlinks, maybe the install code can be part of that function instead?

That would be `link_python_dir`, which runs too early. This was mostly working already, since like `link_python_dir` the install code did not require the python code to be in a `python` subdirectory. However, python code was only being installed if there was a corresponding compiled library, which was incorrect. `link_python_dir` now accumulates a list, similar to the one produced by `add_spt3g_library`, which is used for installation, so this should be fully fixed. 

> For me, "make install" does not copy over spt3g/version.py, so package import fails...

That wasn't just you, and was purely a bug in the cmake script that couldn't be worked around as a user. 

> (I've also tried specifying 'core;version' -- the problem is not that version.py is not built, it's that it is not installed.)

Right, and it isn't a project, so it isn't valid to add to the list. However, since nothing works without it, it should always be installed. 

There were also some problems with the library rpath settings which were causing things to secretly still depend on the build directory contents which should now be fixed, mostly by moving the global default settings to happen before the individual library targets are created, so that they get picked up.

mhasself reviewed (approved):
This works for me!

I agree with earlier comments that installing too-generically named "libcore.so" into /usr/lib might be presumptuous ... libspt3g-core.so and friends would look pretty slick though.

arahlin reviewed (approved):
Looks good to me, would be good to rename the compiled library files to include `spt3g` in the name before merging, now that `load_pybindings` doesn't depend so strictly on the filenames.

cnweaver commented:
> for a brand new build, the installer should byte-compile the python files during the installation process, to make sure the code can be imported properly from a read-only file system. There is some code already in place for that as part of the tarball make target that you could probably copy over.

This is now done. I'm a bit baffled to find that running `python -m compileall -f ${python_dir}` just doesn't seem to do anything on my Mac laptop, but it does the expected thing on Linux, which I think will cover most cases where users are likely to need this. (I'm guessing no one is trying to put a Mac version in CVMFS.)

> Looks good to me, would be good to rename the compiled library files to include spt3g in the name before merging, now that load_pybindings doesn't depend so strictly on the filenames.

I have a patch for this working (I think) locally, but it turns out to be moderately invasive, so I wanted to confirm that people want this before committing it. Specifically, to change the installed library names it was necessary to change the target names, so this has visible effect on all downstream cmake code, namely the individual project `CMakeList.txt`s and `so3g`s build system. Since this change will effectively namespace the libraries anyway, (in the uncommitted patch) I have removed the `spt3g` CMake exported target namespace, so the the only change required on the so3g side is to change `target_link_libraries(so3g spt3g::core)` to `target_link_libraries(so3g spt3g::core)`. In the patch I have updated all of the projects in this repository, but if the change is merged into the internal spt3g-software it will require updating the additional projects which that repository has. 

If all of that sounds acceptable, I can push this as well.

arahlin commented:
Wow I’m really surprised this isn’t as simple as changing the library prefix variable from `lib` to `libspt3g-` or something…

cnweaver commented:
Oh, yes, that does work. It seems kind of underhanded, in a way, but eliminates all of the problems I was concerned about, so let's go with it. 

(I was thinking about this like the traditional `-l` option for `$CC` does, so `lib` is the prefix and `spt3g-` must be part of the name. . . but it's all one string in the end.)

arahlin commented:
Awesome!  No further issues from me.

arahlin commented:
Maybe ignore .pyc files found in the source tree when installing the python directories, so that the byte-compile always rebuilds them using the correct python executable on install?  The byte-compile worked on my mac when I removed all the old .pyc files.

cnweaver commented:
Sorry I got distracted from this for a while. I've added the filter for existing .pyc and .pyo files; although it makes no difference to the oddity I was seeing, it does seem like a generally useful addition.

arahlin commented:
Not sure what the issue is with the mac installer, but I agree that it's not a critical issue.  Seems like this is otherwise ready to merge to me!

nwhitehorn commented on CMakeLists.txt:
```diff
@@ -81,6 +83,39 @@ execute_process(COMMAND mkdir -p ${CMAKE_LIBRARY_OUTPUT_DIRECTORY})
+	list(APPEND BUILD_PROJECTS core)
+endif(${USE_PROJECT_LIST} AND WILL_BUILD_CORE EQUAL -1)
+
+# If not set by the user, ask python when it keeps modules
```
"when" or "where"?

cnweaver replied:
Fixed.

nwhitehorn reviewed (approved):
Looks good to me! (One typo found)
